finality-aleph/src/data_io/data_store.rs

```diff
@@ -309,6 +309,27 @@ where
                             debug!(target: "aleph-data-store", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block, time_waiting.as_secs());
                             self.block_requester
                                 .request_stale_block(block.hash, block.num);
+                        } else {
+                            // The whole branch has been imported, what's holding us must be that the parent of the base is not finalized
+                            let bottom_block = proposal.bottom_block();
+                            if let Ok(parent_hash) =
+                                self.chain_info_provider.get_parent_hash(&bottom_block)
+                            {
+                                let parent_num = bottom_block.num - NumberFor::<B>::one();
+                                if self
+                                    .chain_info_provider
+                                    .get_finalized_at(parent_num)
+                                    .is_err()
+                                {
+                                    debug!(target: "aleph-data-store", "Requesting a justification for block {:?} {:?} after it has been missing for {:?} secs.", parent_num, parent_hash, time_waiting.as_secs());
+                                    self.block_requester
+                                        .request_justification(&parent_hash, parent_num);
+                                } else {
+                                    warn!(target: "aleph-data-store", "The proposal {:?} is pending even though blocks were imported and parent was finalized", proposal);
```

Comment:
could it be possible that even though some block at `parent_num` is finalized it is not the parent of the `bottom_block`?

Reply:
we were requesting parent of `BlockHashNum` so for both height and hash, so I guess we are safe

Reply:
But we only checked if some block at height of `parent_num` is finalized

Reply:
This is possible but only in the presence of malicious nodes (follows from the fact that honest nodes always "root" their proposal in a finalized block). In such a case this proposal will not go through and will stay pending, which is OK. There is a pruning mechanism which will get rid of it if we run low on memory. We could attempt removing it right away, but that would add complexity to the code.

Reply:
I've added more precise logging to make it possible to see in case this happens.